cpp/src/arrow/dataset/file_base.cc

```diff
@@ -47,77 +46,161 @@ Result<ScanTaskIterator> FileDataFragment::Scan(ScanContextPtr context) {
+Result<DataSourcePtr> FileSystemDataSource::Make(
+    fs::FileSystemPtr filesystem, fs::FileStatsVector stats,
+    ExpressionPtr source_partition, const PartitionSchemePtr& partition_scheme,
+    const std::string& partition_base_dir, FileFormatPtr format) {
+  ExpressionVector partitions(stats.size(), scalar(true));
+  ARROW_ASSIGN_OR_RAISE(auto forest, fs::PathForest::Make(std::move(stats)));
+
+  // TODO(bkietz) extract partition_base_dir and partition_scheme back out to discovery
 
-  auto visitor = [&files](const fs::FileStats& stats) {
-    if (stats.IsFile()) {
-      files.emplace_back(new fs::FileStats(stats));
+  // apply partition_scheme to forest to derive partitions
+  auto apply_partition_scheme = [&](fs::PathForest::Ref ref) {
+    if (auto relative =
+            fs::internal::RemoveAncestor(partition_base_dir, ref.stats().path())) {
+      auto segments = fs::internal::SplitAbstractPath(relative->to_string());
+
+      if (segments.size() > 0) {
+        auto segment_index = static_cast<int>(segments.size()) - 1;
+        auto maybe_partition = partition_scheme->Parse(segments.back(), segment_index);
+
+        partitions[ref.i] = std::move(maybe_partition).ValueOr(scalar(true));
+      }
     }
     return Status::OK();
   };
-  // The matcher ensures that directories (and their descendants) are not
-  // visited.
-  auto matcher = [this, options](const fs::FileStats& stats, bool* match) {
-    *match = this->PartitionMatches(stats, options->filter);
-    return Status::OK();
-  };
 
-  for (auto tree : forest_) {
-    DCHECK_OK(tree->Visit(visitor, matcher));
+  DCHECK_OK(forest.Visit(apply_partition_scheme));
+
+  return DataSourcePtr(new FileSystemDataSource(
+      std::move(filesystem), std::move(forest), std::move(source_partition),
+      std::move(partitions), std::move(format)));
+}
+
+std::string FileSystemDataSource::ToString() const {
+  std::string repr = "FileSystemDataSource:";
+
+  if (forest_.size() == 0) {
+    return repr + " []";
   }
 
-  auto file_it = MakeVectorIterator(std::move(files));
-  auto file_to_fragment = [options, this](std::unique_ptr<fs::FileStats> stats,
-                                          std::shared_ptr<DataFragment>* out) {
-    FileSource src(stats->path(), filesystem_.get());
-    ARROW_ASSIGN_OR_RAISE(*out, format_->MakeFragment(src, options));
+  DCHECK_OK(forest_.Visit([&](fs::PathForest::Ref ref) {
+    repr += "\n" + ref.stats().path();
+
+    if (!partitions_[ref.i]->Equals(true)) {
+      repr += ": " + partitions_[ref.i]->ToString();
+    }
+
     return Status::OK();
-  };
+  }));
 
-  return MakeMaybeMapIterator(file_to_fragment, std::move(file_it));
+  return repr;
 }
 
-bool FileSystemDataSource::PartitionMatches(const fs::FileStats& stats,
-                                            ExpressionPtr filter) {
-  if (filter == nullptr) {
-    return true;
+util::optional<std::pair<std::string, std::shared_ptr<Scalar>>> GetKey(
```

Comment:
Partition schemes are more general than key value pairs. Trying to shoehorn a partition representing (for example) a range constraint into a key value pair representation is far more brittle than expression destructuring. (Related: I've introduced the sub interface `PartitionKeysScheme` which *does* explicitly deal in key value pairs.) As with the recent implicit cast issue, I would say that if a problem exists with simplification (currently synonymous with `Expression::Assume`) then that problem should be repaired rather than reimplementing a subset of it elsewhere.

Reply:
To support the special (but common) case of key value partition information, we could have PartitionScheme parse to `variant<ExpressionPtr, pair<string, ScalarPtr>>` (more likely a wrapper). Partition schemes like Hive or Schema which emit key value pairs exclusively could opt to emit pre-unboxed expressions